caput/mpiarray.py

```diff
@@ -89,6 +92,100 @@
             print(ri, arr3)
         mpiutil.barrier()
 
+Direct Slicing
+==============
+
+The :class: `MPIArray` supports direct slicing, with `__getitem__`.
+This can be used for both fetching and assignment. It is recommended
+to only index into the non-parallel axis or to do a full slice `[:]`.
+
+Behaviour
+
+- A full slice `[:]` will return a :class: `MPIArray` on fetching, with
+identical properties to the original array.
+- Any indexing or slicing into the non-parallel axis, will return a
+:class: `MPIArray`. The number associated with the parallel axis,
+will be adjusted if a slice results in an axis reduction.
+- Any indexing into the parallel axis will result into a local index on each rank,
+returning a regular `numpy` array.
+
+Example
+-------
+
+    import numpy as np
+    from caput import mpiarray, mpiutil
+
+    # Direct indexing into parallel axis returns a numpy array
+    # equal to local array indexing
+
+    darr = mpiarray.MPIArray((size,), axis=0)
+    (darr[0] == darr.local_array[0]).all()
+    hasattr(darr[0], "axis") == False
+
+    # indexing into non-parallel axes
+    darr = mpiarray.MPIArray((4, size), axis=1)
+    darr[:] = mpiutil.rank
+
+    darr[0] == mpiutil.rank
+    darr[0].axis == 0
+
+ufunc
+=====
+
+In NumPy, universal functions (or ufuncs) are functions that operate on ndarrays
+in an element-by-element fashion. :class: `MPIArray` supports all ufunc calculations,
+except along the parallel axis.
+
+Requirements
+
+- All input MPIArrays *must* be distributed along the same axis.
+- If you pass a kwarg `axis` to the ufunc, it must not be the parallel axis.
+
+Behaviour
+---------
+
+- If no output are provided, the results are converted back to MPIArrays. The new array
+will either be parallel over the same axis as the input MPIArrays, or possibly one axis
+down if the `ufunc` is applied via a `reduce` method (i.e. the shape of the array is reduced
+by one axis).
+- For operations that normally return a scalar, the scalars will be wrapped into a 1D array,
+parallel across axis 0.
+- shape related attributes will be re-calculated
+
+Example
+------
+
+    import numpy as np
+    from caput import mpiarray, mpiutil
+
+    dist_arr = mpiarray.MPIArray((mpituil.size, 4), axis=0)
```

Comment:
A practical example of why doctest is useful...
```suggestion
    dist_arr = mpiarray.MPIArray((mpiutil.size, 4), axis=0)
```